src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -495,7 +496,13 @@ void FinalizeWeakReference(Object * obj)
 
     WEAKREFERENCEREF pThis((WeakReferenceObject *)(obj));
 
-    OBJECTHANDLE handle = AcquireWeakHandleSpinLock(pThis);
+    // Note:
+    // We need to revisit the implementation of FinalizeWeakReference() in case of concurrent ScanForFinalization
+    // ThreadSuspend::g_pSuspensionThread is set before SuspendRuntime is called, during that time interval, 
+    // there could be managed code running calling WeakReferenceNative::SetTarget() and missed the synchronization.
+    bool isRuntimeSuspended = (ThreadSuspend::GetSuspensionThread() != nullptr);
```

Comment:
@Maoni0 and I were reading about this and we were worried about the possibility described above. Here is a detailed description of what the problem is.

Suppose Thread1 is attempting to suspend the runtime. It reaches the point (any time after the `g_pSuspensionThread` is set and any time before the runtime is completely suspended so that there is no possibility for managed code to run)
https://github.com/dotnet/runtime/blob/f6f4ab624bd3e55afab4262ba9207f3f6329039a/src/coreclr/src/vm/threadsuspend.cpp#L6427-L6465

Thread 2 is running this function, so `isRuntimeSuspended` would be true and we will skip the synchronization for performance. (The could be triggered by background GC)
Thread 3 is running `WeakReferenceNative::SetTarget`, which assumed concurrent access to the underlying handle is synchronized by the spinlock. (If the flag really meant no managed code could be executing, then this couldn't happen, but it isn't)

In the light of this, @kouvel, **do we have a better flag to indicate the runtime is completely suspended?**

To avoid duplicated futile attempts, let me share some experiments and findings I had before.

With more code reading, I found this flag to be potentially useful, yet I concluded that it is not.

https://github.com/dotnet/runtime/blob/f458881ad8c3d4c07a270b9319878402355b99f7/src/coreclr/src/vm/threadsuspend.h#L205-L230

Beside the scary comment, this flag is set inside the implementation of `SuspendRuntime`. The flag is `false` after `g_pSuspensionFlag` is set and before `SuspendRuntime` is called, making it just as unusable, even when we do
`ThreadSuspend::GetSuspensionThread() != nullptr && !ThreadSuspend:: SysIsSuspendInProgress()`.

I also considered `ThreadSuspend::GetSuspensionThread() != nullptr && ThreadSuspend::GetSuspensionThread() == ::GetThread()`. This can never return true if the runtime is not completely suspended, yet this would be wasteful. For server GC, all but one thread calling `FinalizeWeakReference` is managed, so `::GetThread` return `nullptr` for the unmanaged ones and reduce the usefulness of this optimization.